### PR TITLE
feat: Add npu megatron support

### PR DESCRIPTION
**Pull request by UsernameFull**
## Summary
This PR adds support for Huawei Ascend NPU devices with Megatron-Core backend, enabling ROLL framework to run reinforcement learning training on NPU hardware.

### 1. Platform Detection Priority
File: `roll/platforms/__init__.py`

Changes: Reordered platform detection to check NPU before CUDA.

Reason: NPU devices were incorrectly falling back to CUDA platform. Prioritizing NPU detection ensures NpuPlatform is properly initialized when torch_npu is available.

### 2. Device-Agnostic Operations
File: `roll/pipeline/base_worker.py`

Changes:

- Replaced "cuda" with current_platform.device_type
- Replaced torch.cuda.memory_allocated() with current_platform.memory_allocated()

### 3. MindSpeed Integration
File: `mcore_adapter/src/mcore_adapter/training_args.py`

Changes: Added optional import of mindspeed.megatron_adaptor .

Reason: MindSpeed is Huawei's library providing NPU-specific Megatron optimizations. The adaptor patches Megatron-Core for NPU compatibility while maintaining GPU compatibility via try-except.

### 4. NPU Attention Mask Format
File: `roll/distributed/strategy/megatron_strategy.py`

Changes: Added NPU-specific attention mask transformation to 4D format.

Reason: NPU requires 4D attention masks [B, 1, S, S] instead of standard 2D [B, S] . This hardware-specific requirement ensures correct attention computation on NPU.

```
if hasattr(torch, "npu") and torch.npu.is_available() and attention_mask is 
not None:
    attention_mask = attention_mask.bool()
    attention_mask = attention_mask[:, None, None, :].expand(B, 1, S, S)
```
### 5. Optimizer Compatibility
File: `roll/third_party/megatron/optimizer.py`

Changes: Added support for no_weight_decay_cond , scale_lr_cond , lr_mult parameters.

### 6. Example Configurations
Files:
```
- examples/ascend_examples/qwen3_4B_dpo_megatron.yaml
- examples/ascend_examples/qwen3_8b_rlvr_deepspeed.yaml
- examples/ascend_examples/run_dpo_pipeline.sh
```
Reason: Provides ready-to-use NPU training examples demonstrating proper device mapping and strategy configuration for both DPO and RLVR pipelines.

## Impact
Benefits:

- Enables megatron on Huawei Ascend NPU hardware
- Maintains full backward compatibility with GPU systems
- Follows existing platform abstraction patterns

## Requirements
- Huawei Ascend NPU with torch_npu installed
- MindSpeed(v0.15.3) library for NPU Megatron support

### CODE REVIEW

File: roll/platforms/__init__.py [hunk 1]
```diff
@@ -25,26 +25,31 @@ def _init_platform() -> Platform:
     Returns:
         An instance of a subclass of Platform corresponding to the detected hardware.
     """
+    try:
+        import torch_npu  # noqa: F401
+
+        if hasattr(torch, "npu") and torch.npu.is_available():
+            logger.debug("Detected NPU (torch_npu). Initializing NPU platform.")
```
Comment (HuangJoJo) [inline roll/platforms/__init__.py:32]: plz change log_level to info

File: mcore_adapter/src/mcore_adapter/training_args.py [hunk 2]
```diff
@@ -2,6 +2,12 @@
 from dataclasses import dataclass, field, fields
 from typing import Literal, Optional, Union
 
+try:
+    # NPU patch
+    import mindspeed.megatron_adaptor
```
Comment (HuangJoJo) [inline mcore_adapter/src/mcore_adapter/training_args.py:8]: mindspeed will be installed in NPU prebuilt docker？

**Comment by HuangJoJo** [comment]
plz attach qwen3-8B rlvr megatron training curves

File: mcore_adapter/src/mcore_adapter/training_args.py (hunk 2 above)
Comment (UsernameFull) [reply mcore_adapter/src/mcore_adapter/training_args.py:8]: Yes,NPU Docker configuration will be created later, including packages such as torch_npu, vll-ascend, and mindspeed.

**Comment by UsernameFull** [comment]
> plz attach qwen3-8B rlvr megatron training curves

image

File: roll/platforms/__init__.py (hunk 1 above)
Comment (UsernameFull) [reply roll/platforms/__init__.py:32]: The CUDA logs and other logs under this file are also `logger.debug`.
